### PR TITLE
Improve structure visualization

### PR DESCRIPTION
**Pull request by yakutovicha**
Despite the branch is called "structure_editor" I was mostly working on the Structure visualizer here. It had to be improved before the editor can be added.

### CODE REVIEW

File: aiidalab_widgets_base/__init__.py [hunk 1]
```diff
@@ -4,15 +4,17 @@
 from aiida import load_profile
 load_profile()
 
-from .viewers import viewer  # noqa
 from .codes import CodeDropdown, AiiDACodeSetup, valid_aiidacode_args  # noqa
 from .computers import SshComputerSetup, valid_sshcomputer_args  # noqa
 from .computers import AiidaComputerSetup, valid_aiidacomputer_args  # noqa
 from .computers import ComputerDropdown  # noqa
 from .databases import CodQueryWidget  # noqa
+#from .editors import editor  # noqa
```
Comment (csadorf) [inline aiidalab_widgets_base/__init__.py]: Is this related to this PR?

File: aiidalab_widgets_base/__init__.py [hunk 2]
```diff
@@ -4,15 +4,17 @@
...
 from .export import ExportButtonWidget  # noqa
 from .process import ProcessFollowerWidget, ProgressBarWidget, RunningCalcJobOutputWidget, SubmitButtonWidget  # noqa
 from .structures import StructureManagerWidget  # noqa
 from .structures import StructureBrowserWidget, StructureExamplesWidget, StructureUploadWidget, SmilesWidget  # noqa
 from .structures_multi import MultiStructureUploadWidget  # noqa
+from .viewers import viewer  # noqa
+
```
Comment (csadorf) [inline aiidalab_widgets_base/__init__.py]: Why is this pulled in to the root namespace as part of this PR?

File: aiidalab_widgets_base/utils/__init__.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 from __future__ import absolute_import
 from __future__ import print_function
+from six.moves import range
```
Comment (csadorf) [inline aiidalab_widgets_base/utils/__init__.py]: We don't support Py27 anymore so this should not be necessary.

File: aiidalab_widgets_base/viewers.py [hunk 4]
```diff
@@ -1,12 +1,15 @@
 """Jupyter viewers for AiiDA data objects."""
-from __future__ import print_function
```
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py]: Maybe we should remove all py27 relicts in a separate patch?

File: aiidalab_widgets_base/viewers.py [hunk 5]
```diff
@@ -15,12 +18,23 @@ def viewer(obj, downloadable=True, **kwargs):
     :type downloadable: bool
 
     Returns the object itself if the viewer wasn't found."""
+    import inspect
+    import warnings
+    from aiida.orm import Node
+
+    if not (inspect.isclass(type(obj)) and issubclass(type(obj), Node)):  # only working with AiiDA nodes
```
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py]:
```suggestion
    if not isinstance(obj, Node)):  # only working with AiiDA nodes
```

File: aiidalab_widgets_base/viewers.py [hunk 6]
```diff
@@ -15,12 +18,23 @@ def viewer(obj, downloadable=True, **kwargs):
...
 
     try:
         _viewer = AIIDA_VIEWER_MAPPING[obj.node_type]
         return _viewer(obj, downloadable=downloadable, **kwargs)
-    except (AttributeError, KeyError):
-        return obj
+    except (KeyError) as exc:
+        if obj.node_type in str(exc):
+            warnings.warn("Did not find an appropriate viewer for the {} object. Returning the object "
+                          "itself.".format(type(obj)))
```
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py]: At least one of the two warnings appears redundant. Maybe combine the two warnings?

File: aiidalab_widgets_base/viewers.py [hunk 7]
```diff
@@ -71,30 +84,182 @@ class StructureDataViewer(ipw.VBox):
     :param downloadable: If True, add link/button to download the content of the object
     :type downloadable: bool"""
 
-    def __init__(self, structure=None, downloadable=True, **kwargs):
+    SELECTION_SIZE = 10
```
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py]: I think a traitlet or a member attribute would be more appropriate for this.

File: aiidalab_widgets_base/viewers.py [hunk 8]
```diff
@@ -71,30 +84,182 @@ class StructureDataViewer(ipw.VBox):
...
+        self._selected_atoms.observe(self._apply_selection, names='value')
+        self.wrong_syntax = ipw.HTML(
+            value="""<i class="fa fa-times" style="color:red;font-size:2em;" ></i> wrong syntax""",
+            layout={'visibility': 'hidden'})
+        self._selection = set()
+        copy_selection_to_clipboard = ipw.Button(description="Copy to clipboard")
+        clear_selection = ipw.Button(description="Clear selection")
+        clear_selection.on_click(self.clear_selection)
+
+        def copy_to_clipboard(change=None):  # pylint:disable=unused-argument
```
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py]: I'd suggest to move this into a module function, potentially as part of a util module.

File: aiidalab_widgets_base/__init__.py (hunk 1 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/__init__.py]: It is not related, I will add editors in another PR

File: aiidalab_widgets_base/__init__.py (hunk 2 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/__init__.py]: it is not pulled, I just ordered things alphabetically

File: aiidalab_widgets_base/viewers.py (hunk 4 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: Yes, I wanted to do a separate PR just to get rid of all the python2 things. But I think it doesn't harm if I remove some of them here and there.

File: aiidalab_widgets_base/viewers.py (hunk 5 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: this doesn't work with python3.7

File: aiidalab_widgets_base/viewers.py (hunk 6 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: Why? The first warning is about a non-aiida object, the second one tells that for the given AiiDA object the viewer is not implemented

File: aiidalab_widgets_base/utils/__init__.py (hunk 3 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/utils/__init__.py]: ok

File: aiidalab_widgets_base/viewers.py (hunk 8 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: ok, nice idea. Will move it to utils.

File: aiidalab_widgets_base/__init__.py (hunk 2 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/__init__.py]: I mean, I didn't add it just now - I just replaced it.

File: aiidalab_widgets_base/viewers.py (hunk 7 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: I am not sure here. `SELECTION_SIZE` is the size of a ball that appears when you select an atom, I don't think it should be defined per member. Traitlet doesn't make sense since I do not expect to call a function when it gets modified.

**Comment by yakutovicha** [comment]
> I'd suggest to identify a few attribute that would be better implemented as traitlets. The atom selection would come to mind for example. 

I agree. Actually, atom selection it is a traitlet already in nglview. I can add `on_atom_select` function that registers an observer for the atom selection. 

Another thing would be the current frame index.

any more ideas?

File: aiidalab_widgets_base/utils/__init__.py (hunk 3 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/utils/__init__.py]: done

File: aiidalab_widgets_base/viewers.py (hunk 8 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py (hunk 4 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: So do you want me to revet this things back?

File: aiidalab_widgets_base/viewers.py (hunk 5 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: No, you are right. I confused myself. `issubclass` in Python 3.7 expects a class, not an object. `isinstance` works fine.

File: aiidalab_widgets_base/viewers.py (hunk 4 above)
Comment (csadorf) [reply aiidalab_widgets_base/viewers.py]: My suggested approach would be to apply all of the Py27-removal-related changes on a separate branch, we can quickly merge that, and then rebase on master. That would automatically clean up this PR. I am happy to help with that if you like.

Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: Alright, thanks for the proposition. I will then rebase things on top of your changes.

File: aiidalab_widgets_base/viewers.py (hunk 7 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: Ok, I renamed it to avoid confusions.

File: aiidalab_widgets_base/viewers.py [hunk 9]
```diff
@@ -71,30 +84,182 @@ class StructureDataViewer(ipw.VBox):
...
+
+    @property
+    def shortened_selection(self):
+        return " ".join([
+            str(t) if isinstance(t, int) else "{}..{}".format(t[0], t[1]) for t in find_ranges(sorted(self._selection))
+        ])
+
     @property
     def structure(self):
         """Returns ASE Atoms object."""
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: agreed.

File: aiidalab_widgets_base/viewers.py [hunk 10]
```diff
@@ -71,30 +84,182 @@ class StructureDataViewer(ipw.VBox):
...
+            self.wrong_syntax.layout.visibility = 'visible'
+
+    def clear_selection(self, change=None):  # pylint:disable=unused-argument
+        self._viewer._remove_representations_by_name(repr_name='selected_atoms')  # pylint:disable=protected-access
+        self._selection = set()
+        self._selected_atoms.value = self.shortened_selection
+
+    @property
+    def frame(self):
+        return self._viewer.frame
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: agreed

File: aiidalab_widgets_base/viewers.py [hunk 11]
```diff
@@ -71,30 +84,182 @@ class StructureDataViewer(ipw.VBox):
...
+
+    @frame.setter
+    def frame(self, frame):
+        self._viewer.frame = frame
+
+    @property
+    def shortened_selection(self):
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: agreed.

File: aiidalab_widgets_base/viewers.py [hunk 12]
```diff
@@ -110,46 +275,61 @@ def structure(self, structure):
...
-            self._structure = structure
-            self._name = structure.get_chemical_formula()
-        elif isinstance(structure, Node):
-            self._structure = structure.get_ase()
-            self._name = structure.id
-        elif structure is None:
-            self._structure = None
-            self._name = None
-            return  # if no structure provided, the rest of the code can be skipped
+        if isinstance(structure, (list, tuple)):
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: ok. I created two attributes: `structure` and `trajectory`, however I still did not manage to properly synchronize them. @csadorf can you please have a look at my last commit 
602efb2?

**Comment by yakutovicha** [comment]
@csadorf, as I commented [above](https://github.com/aiidalab/aiidalab-widgets-base/pull/47#discussion_r366716221), I've added two traitlets: `structure` and `trajectory`. The implementation can be found [here](https://github.com/aiidalab/aiidalab-widgets-base/blob/602efb21a3d5cb11e9a3f177b37288b7ca6a9a9b/aiidalab_widgets_base/viewers.py#L242-L310).

 However, I still didn't manage to synchronize them properly. What I want to achieve is: if I modify the `structure`, the `trajectory` gets updated and if I modify `trajectory`, the `structure`  gets updated. You can look at the implementation, but in essence, I tried to use [hold_trait_notifications](https://github.com/aiidalab/aiidalab-widgets-base/blob/602efb21a3d5cb11e9a3f177b37288b7ca6a9a9b/aiidalab_widgets_base/viewers.py#L297) assuming that it would prevent functions registered with `observe` method to be called. However, they are still called.

any idea how to achieve that?

File: aiidalab_widgets_base/viewers.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (csadorf) [reply aiidalab_widgets_base/viewers.py]: @yakutovicha Yes, I will have a look at it.

**Comment by yakutovicha** [comment]
@csadorf, I think I found relatively good solution. See the last commit

File: aiidalab_widgets_base/viewers.py [hunk 13]
```diff
@@ -71,30 +84,182 @@ class StructureDataViewer(ipw.VBox):
...
                 description="File format:",
             )
             self.download_btn = ipw.Button(description="Download")
             self.download_btn.on_click(self.download)
             children.append(ipw.HBox([self.file_format, self.download_btn]))
+
+        if 'children' in kwargs:
+            children += kwargs.pop('children')
+
         super().__init__(children, **kwargs)
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 14]
```diff
@@ -71,30 +84,182 @@ class StructureDataViewer(ipw.VBox):
...
+            return
+        self._viewer._remove_representations_by_name(repr_name='selected_atoms')  # pylint:disable=protected-access
+        self._viewer.add_ball_and_stick(  # pylint:disable=no-member
+            name="selected_atoms",
+            selection=vis_list,
+            color=color,
+            aspectRatio=size,
+            opacity=opacity)
+
+    def _apply_selection(self, change=None):  # pylint:disable=unused-argument
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 15]
```diff
@@ -12,12 +19,20 @@ def viewer(obj, downloadable=True, **kwargs):
     :type downloadable: bool
 
     Returns the object itself if the viewer wasn't found."""
+    import warnings
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 16]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
             payload = base64.b64encode(dataf.to_csv(index=False).encode()).decode()
             fname = '{}.csv'.format(parameter.pk)
             to_add = """Download table in csv format: <a download="{filename}"
             href="data:text/csv;base64,{payload}" target="_blank">{title}</a>"""
             self.value += to_add.format(filename=fname, payload=payload, title=fname)
 
 
-class StructureDataViewer(ipw.VBox):
-    """Viewer class for structure object
+class StructureDataBaseViewer(ipw.VBox):
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 17]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
-    :type downloadable: bool"""
+    :param configure_view: If True, configuration tabs
+    :type configure_view: bool"""
+    selection = Set(Int, read_only=True)
+    DEFAULT_SELECTION_OPACITY = 0.2
+    DEFAULT_SELECTION_RADIUS = 6
+    DEFAULT_SELECTION_COLOR = 'green'
+
+    def __init__(self, configure_view=True, **kwargs):
+        from .utils import CopyToClipboardButton
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/utils/__init__.py [hunk 18]
```diff
@@ -37,5 +39,71 @@ def get_ase_from_file(fname):
...
+    for rng in ranges:
+        try:
+            start, end = rng.split('..')
+            singles += [i for i in range(int(start), int(end) + 1)]
+        except ValueError:
+            return set(), False
+    return set(singles), True
+
+
+class CopyToClipboardButton(ipw.Button):
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/utils/__init__.py]: done

File: aiidalab_widgets_base/utils/__init__.py [hunk 19]
```diff
@@ -37,5 +39,71 @@ def get_ase_from_file(fname):
...
+    """Button to copy text to clipboard."""
+
+    def __init__(self, text_provider_function=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/utils/__init__.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 20]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
+class StructureDataBaseViewer(ipw.VBox):
+    """General viewer class for AiiDA structure objects
 
-    :param structure: structure object to be viewed
-    :type structure: StructureData or CifData
-    :param downloadable: If True, add link/button to download the content of the object
-    :type downloadable: bool"""
+    :param configure_view: If True, configuration tabs
+    :type configure_view: bool"""
+    selection = Set(Int, read_only=True)
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: Because selection is done by clicking on an atom with a mouse,  I don't think it makes sense to specify selection programmatically.

File: aiidalab_widgets_base/viewers.py [hunk 21]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
-class StructureDataViewer(ipw.VBox):
-    """Viewer class for structure object
+class StructureDataBaseViewer(ipw.VBox):
+    """General viewer class for AiiDA structure objects
 
-    :param structure: structure object to be viewed
-    :type structure: StructureData or CifData
-    :param downloadable: If True, add link/button to download the content of the object
-    :type downloadable: bool"""
+    :param configure_view: If True, configuration tabs
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 22]
```diff
@@ -1,8 +1,15 @@
 """Jupyter viewers for AiiDA data objects."""
+import base64
 
 import ipywidgets as ipw
 from IPython.display import display
 import nglview
+from ase import Atoms
+
+from traitlets import Instance, Int, observe, Set, Union, validate
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 23]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
             fname = '{}.csv'.format(parameter.pk)
             to_add = """Download table in csv format: <a download="{filename}"
             href="data:text/csv;base64,{payload}" target="_blank">{title}</a>"""
             self.value += to_add.format(filename=fname, payload=payload, title=fname)
 
 
-class StructureDataViewer(ipw.VBox):
-    """Viewer class for structure object
+class StructureDataBaseViewer(ipw.VBox):
+    """General viewer class for AiiDA structure objects
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: it can't

File: aiidalab_widgets_base/viewers.py [hunk 24]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
+        from .utils import CopyToClipboardButton
 
-    def __init__(self, structure=None, downloadable=True, **kwargs):
+        # Defining viewer box.
 
-        self._structure = None
-        self._name = None
+        # 1. Nglviwer
         self._viewer = nglview.NGLWidget()
+        self._viewer.camera = 'orthographic'
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: What would be the benefit of that?

File: aiidalab_widgets_base/viewers.py [hunk 25]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
+            ])
+        ])
+
+        # Defining appearance tab.
+
+        # 1. Choose background color.
+        background_color = ipw.ColorPicker(description="Background")
+
+        def change_background(change):  # Note: change this to lambda when switched to python3.8
+            self._viewer.background = change['new']
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 26]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
+        # 1. Choose download file format.
+        self.file_format = ipw.Dropdown(options=['xyz', 'cif'], layout={"width": "200px"}, description="File format:")
+
+        # 2. Download button.
+        self.download_btn = ipw.Button(description="Download")
+        self.download_btn.on_click(self.download)
+
+        # 3. Screenshot button
+        screenshot = ipw.Button(description="Screenshot", icon='camera')
+        screenshot.on_click(lambda c: self._viewer.download_image())
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 27]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
-            self._structure = structure.get_ase()
-            self._name = structure.id
-        elif structure is None:
-            self._structure = None
-            self._name = None
-            return  # if no structure provided, the rest of the code can be skipped
+        if index not in self.selection:
+            self.selection.add(index)
+        else:
+            self.selection.discard(index)
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: ok

File: aiidalab_widgets_base/viewers.py [hunk 28]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
+            color=color,
+            aspectRatio=size,
+            opacity=opacity)
+
+    def _apply_selection(self, change=None):  # pylint:disable=unused-argument
+        """Apply selection specified in the text area."""
+        short_selection = change['new']
+        expanded_selection, syntax_ok = string_range_to_set(short_selection)
+        self.set_trait('selection', expanded_selection)
+
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 29]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
+        if syntax_ok:
+            self.wrong_syntax.layout.visibility = 'hidden'
+            self.highlight_atoms(self.selection)
         else:
-            raise ValueError("Unsupported type {}, structure must be one of the following types: "
-                             "ASE Atoms object, AiiDA CifData or StructureData.")
+            self.wrong_syntax.layout.visibility = 'visible'
+
+    def clear_selection(self, change=None):  # pylint:disable=unused-argument
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 30]
```diff
@@ -52,101 +67,200 @@ def __init__(self, parameter, downloadable=True, **kwargs):
...
-        self._viewer.add_unitcell()  # pylint: disable=no-member
+    def shortened_selection(self):
+        return " ".join([
+            str(t) if isinstance(t, int) else "{}..{}".format(t[0], t[1]) for t in find_ranges(sorted(self.selection))
+        ])
 
     def download(self, change=None):  # pylint: disable=unused-argument
         """Prepare a structure for downloading."""
-        from IPython.display import Javascript
+        self._download(payload=self._prepare_payload(), filename='structure.' + self.file_format.value)
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: yes, I want to move _downloads into utils or make a stand-alone widget #46. So preparing infrastructure for that.

File: aiidalab_widgets_base/viewers.py [hunk 31]
```diff
@@ -155,6 +269,43 @@ def thumbnail(self):
         return self._prepare_payload(file_format='png')
 
 
+class StructureDataViewer(StructureDataBaseViewer):
+    """Viewer class for AiiDA structure objects
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: added dots to all class docstrings in the file

File: aiidalab_widgets_base/viewers.py [hunk 32]
```diff
@@ -155,6 +269,43 @@ def thumbnail(self):
...
+
+    :param structure: structure object to be viewed
+    :type structure: StructureData or CifData"""
+    structure = Union([Instance(Atoms), Instance(Node)], allow_none=True)
+
+    def __init__(self, structure=None, **kwargs):
+        super().__init__(**kwargs)
+        if structure:
+            self.structure = structure
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 33]
```diff
@@ -155,6 +269,43 @@ def thumbnail(self):
...
+
+    @validate('structure')
+    def _valid_structure(self, change):  # pylint: disable=no-self-use
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: linter alloways compalins about `elif` after `return` statement. But I moved the `None` check on top.

File: aiidalab_widgets_base/viewers.py [hunk 34]
```diff
@@ -155,6 +269,43 @@ def thumbnail(self):
...
+            return structure
+        if isinstance(structure, Node):
+            return structure.get_ase()
+        if structure is None:
+            return None  # if no structure provided, the rest of the code can be skipped
+        raise ValueError("Unsupported type {}, structure must be one of the following types: "
+                         "ASE Atoms object, AiiDA CifData or StructureData.")
+
+    @observe('structure')
+    def _update_structure_view(self, change):  # pylint: disable=unused-argument
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/viewers.py [hunk 35]
```diff
@@ -155,6 +269,43 @@ def thumbnail(self):
...
+        """Update structure view after structure was modified."""
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: done

File: aiidalab_widgets_base/utils/__init__.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/utils/__init__.py]: alright, I made it a 'value' trait

**Comment by yakutovicha** [comment]
@csadorf, please review

File: aiidalab_widgets_base/misc.py [hunk 36]
```diff
@@ -0,0 +1,39 @@
...
+               document.body.appendChild(el);
+               // Select text inside element
+               el.select();
+               // Copy text to clipboard
+               document.execCommand('copy');
+               // Remove temporary element
+               document.body.removeChild(el);
+            }}
+            copyStringToClipboard("{selection}");
+       """.format(selection=self.value))  # For the moment works for Chrome, but doesn't work for Firefox.
```
Comment (csadorf) [inline aiidalab_widgets_base/misc.py]: Maybe we should create an issue for that.

File: aiidalab_widgets_base/viewers.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (csadorf) [reply aiidalab_widgets_base/viewers.py]: I don't really see why we should therefore make it impossible to make a selection programmatically, but I'm ok with leaving as-is for now.

[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py]: good
